src/DotRecast.Core/Buffers/RcRentedArray.cs

```diff
@@ -1,31 +1,96 @@
-using System;
+﻿using System;
 using System.Buffers;
+using System.Collections.Generic;
 using System.Runtime.CompilerServices;
+using System.Threading;
 
 namespace DotRecast.Core.Buffers
 {
     public static class RcRentedArray
     {
+        private sealed class RentIdPool
+        {
+            private int[] _generations;
+            private readonly Queue<int> _freeIds;
+            private int _maxId;
+
+            public RentIdPool(int capacity)
+            {
+                _generations = new int[capacity];
+                _freeIds = new Queue<int>(capacity);
+            }
+
+            internal RentIdGen AcquireId()
+            {
+                if (!_freeIds.TryDequeue(out int id))
+                {
+                    id = _maxId++;
+                    if(_generations.Length <= id)
+                        Array.Resize(ref _generations, _generations.Length << 1);    
+                }
+
+                return new RentIdGen(id, _generations[id]);
+            }
+
+            internal void ReturnId(int id)
+            {
+                _generations[id]++;
+                _freeIds.Enqueue(id);
+            }
+
+            internal int GetGeneration(int id)
+            {
+                return _generations.Length <= id ? 0 : _generations[id];
+            }
+        }
+        
+        public const int START_RENT_ID_POOL_CAPACITY = 16;
+        private static readonly ThreadLocal<RentIdPool> _rentPool = new ThreadLocal<RentIdPool>(() => new RentIdPool(START_RENT_ID_POOL_CAPACITY));
+        
         public static RcRentedArray<T> Rent<T>(int minimumLength)
         {
             var array = ArrayPool<T>.Shared.Rent(minimumLength);
-            return new RcRentedArray<T>(ArrayPool<T>.Shared, array, minimumLength);
+            return new RcRentedArray<T>(ArrayPool<T>.Shared, _rentPool.Value.AcquireId(), array, minimumLength);
+        }
+
+        internal static bool IsDisposed(RentIdGen rentIdGen)
+        {
+            return _rentPool.Value.GetGeneration(rentIdGen.Id) != rentIdGen.Gen;
+        }
+
+        internal static void ReturnId(RentIdGen rentIdGen)
+        {
+            _rentPool.Value.ReturnId(rentIdGen.Id);
         }
     }
 
-    public class RcRentedArray<T> : IDisposable
+    public readonly struct RentIdGen
+    {
+        public readonly int Id;
+        public readonly int Gen;
+
+        public RentIdGen(int id, int gen)
+        {
+            Id = id;
+            Gen = gen;
+        }
+    }
+
+    public struct RcRentedArray<T> : IDisposable
     {
         private ArrayPool<T> _owner;
         private T[] _array;
+        private readonly RentIdGen _rentIdGen;
```

Comment:
This is tricky one. Since RcRentedArray now a value type, it will copy values. That means when we dispose it, all copies will remain reference to owner and array.
To fix that we introduce rent id. When rent array, get vacant id and generation. On dispose check the generation. If generation is different, that means this rent was already freed. Otherwise release array and increase the generation.